### PR TITLE
feat(eslint-plugin): [no-unsafe-call][no-unsafe-member-access] improve report messages for `this` for `noImplicitThis`

### LINKED ISSUE #3197: [no-unsafe-call] and [no-unsafe-member-access] improve reporting message when `noImplicitThis` is not enabled
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

context: https://github.com/sveltejs/eslint-plugin-svelte3/issues/102#issuecomment-801572731

reproduction: https://github.com/frederikhors/issueWithThisAny

```json
{
  "rules": {
    "@typescript-eslint/no-unsafe-call": 2,
    "@typescript-eslint/no-unsafe-member-access": 2
  }
}
```

```ts
export const store = () => {
  const methods = {
    OnInit() {
      this.Login();
    },

    Login() {
      //
    },

    async Logout() {
      //
    },
  };

  return methods;
};
```

```jsonc
{
  "extends": "@tsconfig/svelte/tsconfig.json",

  "compilerOptions": {
    "module": "ES2020",
    "target": "ES2020",
    "types": ["svelte", "node"],
    // "noImplicitThis": true
  },

  "include": ["*"],
  "exclude": ["node_modules/*", "__sapper__/*", "public/*"]
}
```

**Expected Result**

Reporting with message that tells the user `this` is type `any` because `noImplicitThis` is not enabled

**Actual Result**

No extra tips

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.17.0` |
| `@typescript-eslint/parser`        | `4.17.0` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `7.22.0` |
| `node`                             | `14.16.0` |

### PR DESCRIPTION
**Pull request by JounQin**
close #3197

### CODE REVIEW

**Review by bradzacher — CHANGES_REQUESTED** [review]
we can actually tell if the config option is turned on!
We should only report this message if the option is turned off

Here's an example of another rule checking if `strictNullChecks` is turned on:
https://github.com/typescript-eslint/typescript-eslint/blob/9273441f7592b52620e10432cb2dd4dc5c3b4db1/packages/eslint-plugin/src/rules/strict-boolean-expressions.ts#L114-L117

---

I think if we're going explicitly display this then we should be sure that `this` is the cause and not something else.

eg we should only display the specific message if:
- the "thing" is a member expression on `any`
- `noImplicitThis` is turned off
- `this` is of type `any`

If any of these conditions is false - then we should use the normal message as we can tell that `noImplicitThis` is not the cause.

**Comment by JounQin** [comment]
@bradzacher Thanks! I'll add the option check. And the other two conditions has been checked, isn't it?

**Comment by bradzacher** [comment]
Depends on the rule.

For example, no unsafe call - `this.foo()` will report if EITHER `this` or `this.foo` are `any`

And this only compounds the more members you add ofc.

For no unsafe member - `this.foo.bar` will again report if `this` or `this.foo` are `any`.

**Comment by JounQin** [comment]
I'm still confused, I've already added checking whether the caller/member parent is `this` at https://github.com/typescript-eslint/typescript-eslint/pull/3199/files#diff-57d22a396b83ef3e522e1cb9ce9ed841ebc305e7978abaf79996180f83a5dc87R54-R55 and https://github.com/typescript-eslint/typescript-eslint/pull/3199/files#diff-95b4d02562218eb4ed05f07944d416c25edb4861ac3b215e31db7ca924c97920R73 in this PR, what should I check further?

**Comment by bradzacher** [comment]
you did check if it was an expression around `this`.

For `no-unsafe-call`, it will error if the callee is `any`.

Which means that it will error in this case: `this.foo()`. But what if `this` is well defined, but `foo` is `any`?
i.e. this case:

```ts
class Foo {
  foo: any;

  method() {
    this.foo();
  }
}
```

In this case saying "try turning on `noImplicitAny`" is wrong - it's not the `this` that's `any` - it's `this.foo`.

OTOH, it is correct to say "try turning on `noImplicitAny`" for this case:
```
const Foo = {
  method() {
    this.foo(); // `this` is `any` with `noImplicitAny` turned off
  },
}
```

For `no-unsafe-member-access`, it's *similar*.

Eg imagine `this.foo.bar` - if `this` is `any`, then it makes sense to say "try turning on `noImplicitAny`", but if `this.foo` is `any`, then it doesn't make sense to say that.


Does that make sense?

**Comment by JounQin** [comment]
OK, maybe I get your point now. `this` type should be `any` for the specific messages, right?

**Comment by bradzacher** [comment]
> this type should be any for the specific messages, right?

Correct!

To triple clarify for you:

```ts
// to save me copy-pasting these strings, the messageIds:
//
// - unsafeCall = 'Unsafe call of an `any` typed value.';
// - unsafeCallThis = [
//    'Unsafe call of an `any` typed value. `this` is typed as `any`.',
//    'You can try to fix this by turning on the `noImplicitThis` compiler option, or adding a `this` parameter to the function.',
//   ].join('\n');
//
// - unsafeMemberExpression = 'Unsafe member access {{property}} on an `any` value.'
// - unsafeThisMemberExpression = [
//    'Unsafe member access {{property}} on an `any` value. `this` is typed as `any`.',
//    'You can try to fix this by turning on the `noImplicitThis` compiler option, or adding a `this` parameter to the function.',
//   ].join('\n');



class Foo {
  bar: any;

  method() {
    this.bar();      // no-unsafe-call: unsafeCall

    this.bar.baz()   // no-unsafe-call: unsafeCall
                     // no-unsafe-member-access: unsafeMemberExpression {property: baz}

    this.bar.baz.bam // no-unsafe-member-access: unsafeMemberExpression {property: baz}
  }
}

// noImplicitThis: false
const Obj = {
  method() {
    this.bar();      // no-unsafe-call: unsafeCallThis
                     // no-unsafe-member-access: unsafeThisMemberExpression {property: bar}

    this.bar.baz()   // no-unsafe-call: unsafeCallThis
                     // no-unsafe-member-access: unsafeThisMemberExpression {property: bar}

    this.bar.baz.bam // no-unsafe-member-access: unsafeThisMemberExpression {property: bar}
  },
};
```

With `noImplicitThis: true`, the last example would OFC be full of TypeScript errors, so we don't need to specifically worry about how we report on it.

There is this one special case which would exist with `noImplicitThis: true`, but it's pretty stupid and very much a user-error - so I don't think we need to handle it with a custom message:
```ts
function anyThisFn(this: any) {
  this.bar();      // no-unsafe-call: unsafeCallThis
                   // no-unsafe-member-access: unsafeThisMemberExpression {property: bar}

  this.bar.baz()   // no-unsafe-call: unsafeCallThis
                   // no-unsafe-member-access: unsafeThisMemberExpression {property: bar}

  this.bar.baz.bam // no-unsafe-member-access: unsafeThisMemberExpression {property: bar}
};
```

**Comment by JounQin** [comment]
@bradzacher How about current implementation?

File: packages/eslint-plugin/src/util/getThisExpression.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  while (node) {
+    if (node.type === AST_NODE_TYPES.CallExpression) {
+      node = node.callee;
+    }
+
+    if (node.type === AST_NODE_TYPES.ThisExpression) {
+      return node;
+    }
+
+    if ('object' in node) {
```
Comment (bradzacher) [inline packages/eslint-plugin/src/util/getThisExpression.ts:18]: couldn't this just be `node.type === AST_NODE_TYPES.MemberExpression`?

File: packages/eslint-plugin/src/util/getThisExpression.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      node = node.object;
+    } else {
+      break;
+    }
```
Comment (bradzacher) [inline packages/eslint-plugin/src/util/getThisExpression.ts:22]: Just so we only ever handle one node per iteration - makes this more predictable

```suggestion
    if (node.type === AST_NODE_TYPES.CallExpression) {
      node = node.callee;
    } else if (node.type === AST_NODE_TYPES.ThisExpression) {
      return node;
    } else if ('object' in node) {
      node = node.object;
    } else {
      break;
    }
```

File: packages/eslint-plugin/src/util/getThisExpression.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+export function getThisExpression(
```
Comment (bradzacher) [inline packages/eslint-plugin/src/util/getThisExpression.ts:6]: one case this doesn't handle is short-circuited optional chains which use the `ChainExpression` node
`(this.x?.y).z`

**Review by bradzacher — CHANGES_REQUESTED** [review]
almost there! Just a few comments.
Thanks for your work on this!

File: packages/eslint-plugin/src/util/getThisExpression.ts (hunk 3 above)
Comment (JounQin) [reply packages/eslint-plugin/src/util/getThisExpression.ts:6]: Done!

**Review by bradzacher — APPROVED** [review]
LGTM - thanks for your contribution!

**Pull request merged**